rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,346 @@
+  * `ackId`’s from incoming requests are used to index into
+    `ack_ids_ack_status`. We return true/false depending on the value in the
+    bitvec.
+  * If we return true for an ackId, we mark `ack_ids_in_use[ack_id] = 0` and
+    `ack_ids_ack_status[ack_id] = 0`. The `ackId` can then be reused.
+
+## Proposal: `splunk_hec` Sinks Indexer Acknowledgement
+
+### User Experience
+
+The `splunk_hec` sinks will automatically integrate with indexer
+acknowledgements if the user has enabled it in their Splunk instance (if
+`ackId`’s are present in HEC responses). If indexer acknowledgement is disabled,
+`splunk_hec` sinks will continue to finalize events based on the HEC response
+status code.
+
+### Implementation
+
+* [Splunk recommendations for client integration with indexer acknowledgement](https://docs.splunk.com/Documentation/Splunk/8.2.3/Data/AboutHECIDXAck#Indexer_acknowledgment_client_behavior)
+
+We generate a single `guid` to use as a channel ID and include this in HEC
+requests. After submitting events to HEC, we will parse the HTTP response for an
+`ackId`. If no `ackId` is found, we rely on the current behavior of setting
+`EventStatus` based solely on the HTTP status code.
+
+If an `ackId` is found, we store it in a `Map<u64, bool>` shared with a
+background tokio task which, for all pending `ackId`’s, will query
+`/services/collector/ack`. This background task will query at an interval of 10
+seconds (as recommended by Splunk) and update the shared structure based on the
+results. Meanwhile, back in the HEC response handler, we’ll check the shared
+structure for `true` at 10 second intervals up to a maximum time limit (5
+minutes as recommended by Splunk).
+
+```{rust}
+fn call(&mut self, req: HecRequest) -> Self::Future {
+        let mut http_service = self.batch_service.clone();
+        Box::pin(async move {
+            ...
+            // handle response
+            let response = http_service.call(req).await?;
+            // if ack_id is found
+            self.ack_id_to_status_map.insert(ack_id, false);
+            let mut interval = time::interval(Duration::from_secs(10));
+            let mut retries = 0;
+            while retries < RETRY_LIMIT {
+                interval.tick().await;
+                if self.ack_id_to_status_map.get(ack_id) {
+                    // set EventStatus::Delivered and update map
+                    break;
+                }
+                retries += 1;
+            }
+
+            // if ack_id is not found, fall back on current behavior
+            let event_status = if response.status().is_success() {
+                EventStatus::Delivered
+            } else if response.status().is_server_error() {
+                EventStatus::Errored
+            } else {
+                EventStatus::Failed
+            };
+
+            ...
+        })
+    }
+```
+
+If we receive a `true` response, we remove the ackId from the shared structure
+and set `EventStatus::Delivered`.
```

Comment:
It might be worth mentioning how you intend to manage shared access, ie mutex vs atomic operations.

Reply:
Sure, I will update with some more specific details. I had an `Arc<Mutex<Map>>` in mind.